index.js

```diff
@@ -1061,25 +1134,27 @@ function nested (laterCode, name, key, schema, externalSchema, fullSchema, subKe
       break
     case 'object':
       funcName = asFuncName(name + key + subKey)
-      laterCode = buildObject(schema, laterCode, funcName, externalSchema, fullSchema)
+      laterCode = buildObject(location, laterCode, funcName)
       code += `
         json += ${funcName}(obj${accessor})
       `
       break
     case 'array':
       funcName = asFuncName('$arr' + name + key + subKey) // eslint-disable-line
-      laterCode = buildArray(schema, laterCode, funcName, externalSchema, fullSchema)
+      laterCode = buildArray(location, laterCode, funcName)
       code += `
         json += ${funcName}(obj${accessor})
       `
       break
     case undefined:
       if ('anyOf' in schema) {
-        dereferenceOfRefs(schema, externalSchema, fullSchema, 'anyOf')
-        schema.anyOf.forEach((s, index) => {
-          var nestedResult = nested(laterCode, name, key, s, externalSchema, fullSchema, subKey !== '' ? subKey : 'i' + index, isArray)
+        // beware: dereferenceOfRefs has side effects and changes schema.anyOf
+        var anyOfLocations = dereferenceOfRefs(location, 'anyOf')
+        anyOfLocations.forEach((location, index) => {
```

Comment:
This was by far the most complicated part to understand and fix.

The reason is that it heavily relies on side effects: 
- `dereferenceOfRefs` changes `schema.anyOf`
- calls to `buildCode` also change the schema (replacing any property with a reference to the actual schema, see here: https://github.com/fastify/fast-json-stringify/blob/master/index.js#L631)

This is necessary because ajv will need to have the "dereferenced" schema  at run time to work properly.

My problem was that changing just the schema would loose the full location/context and tests with multiple external schema in anyOf wouldn't work (this would fail: https://github.com/fastify/fast-json-stringify/pull/239/files#diff-2dbacb6426a8e9d68e93312ea682b74dR400).

I think the best solution would have been to dereference the entire schema at the start and avoid dereferencing it piece by piece - but that would require a complete rewrite.

Here I opted for an hack: `dereferenceOfRefs` not only changes `schema.anyOf`, but also returns an array of locations which contain the context about where the dereferenced schema was found.

I also had a different implementation where I was not calling dereferenceOfRefs but was replacing the schema in place:

```js
// var locations = dereferenceOfRefs(location, 'anyOf')
schema.anyOf.forEach((s, index) => {
  var sLocation = mergeLocation(location, { schema: s })
  if (s.$ref) {
    sLocation = refFinder(s.$ref, sLocation)
    s = sLocation.schema
  }

  schema.anyOf[index] = s

  var nestedResult = nested(laterCode, name, key, sLocation, subKey !== '' ? subKey : 'i' + index, isArray)

  code += `
    ${index === 0 ? 'if' : 'else if'}(ajv.validate(${require('util').inspect(s, { depth: null, maxArrayLength: null })}, obj${accessor}))
      ${nestedResult.code}
  `
  laterCode = nestedResult.laterCode
})
```

In the end I've discarded it because it seemed even more unreadable the the one proposed in this PR.

All of this applies equally to `oneOf`

Reply:
Can you:

1. add some of that explanation in the comment before `deferenceOfRefs`?
2. create a new issue explaining this problem and the full solution.

Reply:
Added expanded comment right before the call of `ajv.validate`: https://github.com/fastify/fast-json-stringify/pull/239/files#diff-168726dbe96b3ce427e7fedce31bb0bcR1156-R1162

Reply:
And I've also created the issue explaining the problem: https://github.com/fastify/fast-json-stringify/issues/241

As a bonus I also added this possible bug: https://github.com/fastify/fast-json-stringify/issues/242